enhancements/namespace-auto-pruning.md

```diff
@@ -0,0 +1,170 @@
+---
+title: Namespace Auto-pruning Policies
+authors:
+  - "@bcaton85"
+reviewers:
+  - "@dmage"
+  - "@Sunandadadi"
+approvers:
+  - "@dmage"
+  - "@Sunandadadi"
+creation-date: 2023-08-15
+last-updated: 2023-08-15
+status: provisional
+---
+
+# Namespace Auto-pruning Policies
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions [optional]
+
+1. Should there be a dry run feature to test the policy?
+1. Should this only be implemented in the new UI and not the old UI?
+1. Can a user configure multiple policies per namespace?
+1. Can a user specify tags or repositories by pattern?
+1. Do we want the ability to have tiered priority for auto-prune execution?
+
+## Summary
+
+In order for a container registry to run cost effectively unused resources should be deleted to reclaim storage space. This task should be done automatically in Quay by deleting unused tags through a given policy. The policy used should be configured by the user and a variety policy options should be given. Examples of initial policies would be prune tags older than X number of days or prune tags when more than 100 are present.
+
+## Motivation
+
+Reclaiming space is essential in running a cost efficient registry. Having this done automatically within Quay removes the need for manual intervention or custom software.
+
+### Goals
+
+* User is able to provide a time based policy for auto-pruning tags within a namespace
+* User is able to provide a tag number based policy for auto-pruning tags within a namespace
+* User is able to specify whether tags go into the time machine or are immediately deleted
+* The correct tags are deleted when the policy is applied
+* Pruned tags are tracked via usage logs
+
+### Non-Goals
+
+* Allowing users to specify multiple policies per namespace
+* Applying policies based on repo or tag name pattern matching
+* Allowing policy configurations at the repository level
+
+## Proposal
+
+The details in this proposal is only for namespace auto-pruning policies but designed in a way to support repository auto-pruning policies in the future. This proposal is structured in three parts: how policies are managed, how the policies are structured in the DB, and how the policies will be executed via the auto-prune worker.
```

Comment:
Why did you decide to go for a namespace granularity vs repository when pruning tags? Were there any implementations issue you saw with the latter?

If you can do repository, you could in theory do namespace by going through that all of that namespace's repositories.

Reply:
Great job on outlining the goals of the auto-pruning feature in this pull request! Also, your breakdown into MVP, short-term, and long-term goals provides a clear and structured view of the feature's development trajectory.

"MVP
- Single namespace policies
    - Includes ground work for policy UI, API, and worker

Short term
- Single repository policies
- Multiple namespace policies
- Multiple repository policies

Long term
- Repository name pattern matching for namespace policies
- Tag name pattern matching for repository policies
- Prune untagged artifacts
- System wide auto-pruning configuration
- Dry run policies"

[
] Full roadmap from PM's perspective (https://docs.google.com/document/d/1qA6nxlGmkbVnmA5NtCRYPQERhZlRFEl3LjaODlldbec/edit)

The MVP phase focusing on single namespace policies is a solid starting point. The inclusion of groundwork for policy UI, API, and the worker is a crucial foundation that sets the stage for further development.

Your short-term goals, which include extending to single repository policies and accommodating multiple namespace and repository policies, show a path to the feature's scalability and the potential for varied user scenarios.

The long-term vision is encompassing advanced capabilities like repository name and tag name pattern matching, pruning untagged artifacts, system-wide configuration, and dry run policies. These objectives indicate a commitment to offering a comprehensive and versatile auto-pruning solution over time.

Overall, your pull request presents a well-structured and forward-thinking plan that aligns with the user needs and the feature's growth potential. Your roadmap covers a range of scenarios and considerations, making it a robust foundation for the auto-pruning feature's development. Well done!

Reply:
I think it makes sense to add this MVP / Short-Term / Long-term perspectives to the proposal. Long-term we also need policies based of "recent pull date:

Reply:
Agreed, added